src/main/java/com/mongodb/kafka/connect/sink/dlq/AnalyzedBatchFailedWithBulkWriteException.java

```diff
@@ -0,0 +1,140 @@
+import com.mongodb.lang.Nullable;
+
+import com.mongodb.kafka.connect.sink.StartedMongoSinkTask;
+
+/**
+ * This class allows to conveniently {@linkplain #log() log} and {@linkplain #report() report} to
+ * the DLQ a {@linkplain MongoCollection#bulkWrite(List, BulkWriteOptions) batch} failed with {@link
+ * MongoBulkWriteException}.
+ */
+public final class AnalyzedBatchFailedWithBulkWriteException {
+  private final List<SinkRecord> batch;
+  private final MongoBulkWriteException e;
+  private final BiConsumer<SinkRecord, Exception> errorReporter;
+  private final Map<Integer, Map.Entry<SinkRecord, WriteException>> recordsFailedWithWriteError =
+      new HashMap<>();
+  private final Map<Integer, SinkRecord> recordsFailedWithWriteConcernError = new HashMap<>();
+  private final Map<Integer, SinkRecord> skippedRecords = new HashMap<>();
+  @Nullable private final WriteConcernException writeConcernException;
+  private final WriteSkippedException writeSkippedException = new WriteSkippedException();
+
+  public AnalyzedBatchFailedWithBulkWriteException(
+      final List<SinkRecord> batch,
+      final MongoBulkWriteException e,
+      final BiConsumer<SinkRecord, Exception> errorReporter) {
+    this.batch = batch;
+    this.e = e;
+    this.errorReporter = errorReporter;
+    WriteConcernError writeConcernError = e.getWriteConcernError();
+    writeConcernException =
+        writeConcernError == null ? null : new WriteConcernException(writeConcernError);
+    analyze();
+  }
+
+  private void analyze() {
+    List<BulkWriteError> writeErrors = e.getWriteErrors();
+    WriteConcernError writeConcernError = e.getWriteConcernError();
+    if (writeErrors.isEmpty()) {
+      assertNotNull(writeConcernError);
+      for (int i = 0; i < batch.size(); i++) {
+        recordsFailedWithWriteConcernError.put(i, batch.get(i));
+      }
+    } else {
+      assertTrue(writeErrors.size() == 1);
+      BulkWriteError writeError = writeErrors.get(0);
+      int writeErrorIdx = writeError.getIndex();
+      recordsFailedWithWriteError.put(
+          writeErrorIdx,
+          new AbstractMap.SimpleImmutableEntry<>(
+              batch.get(writeErrorIdx), new WriteException(writeError)));
+      for (int i = writeErrorIdx + 1; i < batch.size(); i++) {
+        skippedRecords.put(i, batch.get(i));
+      }
+      if (writeConcernError != null) {
+        for (int i = 0; i < writeErrorIdx; i++) {
+          recordsFailedWithWriteConcernError.put(i, batch.get(i));
+        }
+      }
+    }
+  }
+
+  public void log() {
+    if (!recordsFailedWithWriteError.isEmpty()) {
+      recordsFailedWithWriteError.forEach(
+          (i, recordFailedWithWriteError) ->
+              StartedMongoSinkTask.log(
+                  singletonList(recordFailedWithWriteError.getKey()),
+                  recordFailedWithWriteError.getValue()));
+    }
+    if (!recordsFailedWithWriteConcernError.isEmpty()) {
+      StartedMongoSinkTask.log(
```

Comment:
Suggestion: decouple the logging mechanism from StartedMongoSinkTask  - pass in a ` BiConsumer<Collection<SinkRecord>, RuntimeException>` to handle the logging.

Reply:
Done.